### ISSUE TITLE
Include `examples` in set-up of `porepy`

### ISSUE THREAD

**Issue by vlipovac** [pmgbergen/porepy#1113]
As of now we the repository structure is given by

```
├── repository
│   ├── src
│   │   ├── porepy/*.py
│   ├── examples
│   │   ├── __init__.py
│   │   ├── *.py
│   ├── setup.py
```

`examples` is currently available up as a second package (not sub-package of `porepy`, but depending on it).

The goal is to have it included in the installation of `porepy`, s.t. it will appear as a module `porepy.examples` when importing `porepy`.

```[tasklist]
### Tasks
- [x] Read up on Python's `setuptools` to discover a "correct" way.
- [ ] If 1. not fruitful, include the import of `examples` explicitly in `porepy.__ini__.py`. (operate on `PYTHONPATH`)
```

In any way, the installation should be such that the default way to install porepy (`pip install .`) works as expected.

**Comment by vlipovac** [comment]
@keileg 
Please take a look at https://github.com/pmgbergen/porepy/tree/setup-examples

I found a solution using `setuptools.setyp` with a full map between packages and directories [see here.](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html)

**2 very important things:**

1. For some reason, `pip install -e .` does not do the trick. One has to make a full install (including building the wheel for porepy) by `pip install .`
2. The package `examples` is not in the directory of `porepy` when imported in an interpreter, since this requires a modification of `porepy/__init__.py`. I.e. one cannot access the examples by
```python
import porepy
print(porepy.examples)  # raises Error, not a member
```
but by
```python
from porepy import examples as ppe
# This will print the content of the examples. I have included some classes in examples.__init__.py so that this is not empty
print(dir(ppe))  
```

To see how this works, checkout out above branch. Delete all installations or porepy and start with `pip install .` in your project directory.

**Comment by vlipovac** [comment]
Point 1, in the previous comment might be related to another issue.
pip gives a warning if executed while being the root user (currently default in the porepy dev container).
It claims being in conflict with the system package manager.
we might consider switching to create a default non-root user in the dev container or setting up a virtual environment.

I also noticed that after installing porepy using pip in the current dev container, and uninstalling it, the uninstall fails because it "can't find the files". pip freeze lists porepy continuously as installed and it is indeed present importable when starting the interpreter.

**Comment by keileg** [comment]
> It claims being in conflict with the system package manager.
> we might consider switching to create a default non-root user in the dev container or setting up a virtual environment.

I agree, we should fix this (though it is a different issue indeed)

**Comment by keileg** [comment]
@IvarStefansson @OmarDuran: See @vlipovac's comments [above ](https://github.com/pmgbergen/porepy/issues/1113#issuecomment-1910221525), in particular his point number 2. Do you think this an acceptable solution? 

@vlipovac: If we are to make the examples directly importable, would you do that with the setup solution, or would you just include it in the top-level `__init__.py`, as you indicated a few days ago?

**Comment by IvarStefansson** [comment]
I don't get what you refer to as Veljko's point number 2.

**Comment by keileg** [comment]
Sorry. See comments regarding the form import will take (in Veljko's comment above)

**Comment by IvarStefansson** [comment]
I still don't get exactly what you're referring to. Shall we discuss in person?

**Comment by vlipovac** [comment]
> @vlipovac: If we are to make the examples directly importable, would you do that with the setup solution, or would you just include it in the top-level `__init__.py`, as you indicated a few days ago?

**TLDR**
We should move `examples/` into `src/porepy/examples` to avoid the complex setup solution.

But we should **not** have it imported in `porepy/__init__.py`.
I.e., default building options will register `examples` as a sub-package but the user has to import them explicitly `from porepy import examples`.

This should be done with `tests` and experimental features (if we have any) as well.

It is a solution which makes life easier for future development and fits the standard, but requires a bit more work now (mypy ..).

### setup, build and distribution

The most essential point is found in the documentation of the [`setup` keywords](https://setuptools.pypa.io/en/latest/references/keywords.html), in the section for `package_dir`:
![image](https://github.com/pmgbergen/porepy/assets/68282540/4e90f72e-5eda-4419-a09d-d816a9683402)

They essentially recommend for packaging, building and distributing, that the structure of source files and directories should reflect the structure of the final python module. This would require moving `examples/` into `src/porepy/examples/`, if we want `examples` to be included in the PorePy package (especially in `porepy/__init__.py`).

My suggestion from two days ago is a workaround and not "good practice" according to this.

`examples` and `tests` are an essential part of public code. But `porepy` is a toolbox and `examples` and `tests` are an application of that toolbox. _You do not need them loaded into the interpreter to be able to work with porepy._

This practice is also found when dealing with (sub-) packages which depend on other (sub-) packages, for technical reasons (isort, circular import).

[Numpy for example](https://github.com/numpy/numpy/tree/main/numpy), has tests in their single source directory but
```python
import numpy
print(numpy.tests)  # error
from numpy import tests 
print(tests)  # no error
```
This practice is also used for experimental features and helps with the dependencies inside the main package.

### Another note

I read somewhere that developers are going away from `setup.py` (which allows the kind of dynamic packaging demonstrated in the branch), and more towards `setup.cfg` or `pyproject.toml`.
`python -m  build` (producing wheels and sdist for uploading porepy to PyPI) does not work with `setup.py` or with the `setup.cfg` as we have it now.
`toml` and `cfg` files need all sub-packages listed explicitly.

**Comment by keileg** [comment]
@vlipovac: Did you receive the necessary clarifications regarding this matter the other day, or do you need more from me/us?
